File: app/src/main/java/com/sofrosyn/a6gnlsolarcalculator/fragments/InverterRatingFragment.java
package com.sofrosyn.a6gnlsolarcalculator.fragments;

import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;

import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textview.MaterialTextView;
import com.sofrosyn.a6gnlsolarcalculator.R;
import com.sofrosyn.a6gnlsolarcalculator.Utils.StateManager;
import com.sofrosyn.a6gnlsolarcalculator.logic.Calculator;

import java.util.List;

import me.itangqi.waveloadingview.WaveLoadingView;


public class InverterRatingFragment extends Fragment {


    private TextInputEditText inverterLoad;
    private TextInputEditText batteryNumber;
    private MaterialTextView estInverterRating;

    private WaveLoadingView waveLoadingView;
    private ChipGroup batteryChips;
    private ChipGroup inverterVoltageChip;
    private View view;
    private MaterialButton inverterNext;
    private MaterialButton calculate;
    private static int batteryAmp = 100;
    private static int invertVoltage = 12;
    private static int divisor = 1;


    private static double ratingKva;
    private static double time2Charge;
    private static double backUpTime;


    public InverterRatingFragment() {



    }




    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        view = inflater.inflate(R.layout.fragment_inverter_rating, container, false);
         initViews();
         initActions();
         return view;
    }



    private void initViews(){


        inverterLoad = view.findViewById(R.id.fragment_inverter_load);
       // batteryLoad = view.findViewById(R.id.fragment_battery_load);

        estInverterRating = view.findViewById(R.id.fragment_inverter_load_text);
        inverterVoltageChip = view.findViewById(R.id.inverter_voltage);
        inverterNext = view.findViewById(R.id.inverter_next);
        batteryNumber = view.findViewById(R.id.fragment_inverter_battery_no);
        calculate = view.findViewById(R.id.inverter_calculate);
        waveLoadingView =view.findViewById(R.id.waveLoadingView);
        batteryChips = view.findViewById(R.id.chips_battery);

    }


    private void initActions(){


        inverterLoad.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

                if(start > 1){
                ratingKva  =  Calculator.loadInKva(Double.parseDouble(String.valueOf(s)));
                estInverterRating.setText(String.format("Your Estimated Inverter Rating is [%sKva] ",ratingKva));
                }else {
                    estInverterRating.setText("Your Estimated Inverter Rating  [Kva] ");
                    waveLoadingView.setCenterTitle("Load");
                }

            }

            @Override
            public void afterTextChanged(Editable s) {
                waveLoadingView.setProgressValue((int) ratingKva* 10);
                estInverterRating.setTextColor(getResources().getColor(R.color.colorPrimaryDark));
                waveLoadingView.setCenterTitle("Inverter: "+ratingKva+"Kva");

            }
        });



        inverterNext.setOnClickListener(v -> {
            int load = Integer.parseInt(inverterLoad.getText().toString());
            int batteryCount = Integer.parseInt(batteryNumber.getText().toString());


            StateManager.getInstance(getActivity())
                    .Inverter(load,(float) ratingKva,(float) backUpTime,invertVoltage,(float)time2Charge,batteryAmp,invertVoltage,batteryCount);
            Navigation.findNavController(view).navigate(R.id.action_inverterRatingFragment_to_batteryFragment);

        });



        batteryChips.setOnCheckedChangeListener((group, checkedId) -> {
    Chip chip = group.findViewById(checkedId);
     if(chip != null){
         batteryAmp = Integer.parseInt(chip.getText().toString());
         Toast.makeText(getActivity(), "Battery Amp: "+batteryAmp, Toast.LENGTH_SHORT).show();
     } });



        inverterVoltageChip.setOnCheckedChangeListener((group, checkedId) -> {
            Chip chip = group.findViewById(checkedId);
            if(chip != null){
                invertVoltage = Integer.parseInt(chip.getText().toString());
                Toast.makeText(getActivity(), "Inverter Voltage: "+invertVoltage, Toast.LENGTH_SHORT).show();

                switch (invertVoltage){
                    case 12:
                        divisor = 1;
                        Log.d("Divisor", " "+ divisor);
                        break;
                    case 24:
                        divisor = 2;
                        Log.d("Divisor", " "+ divisor);
                        break;
                    case 48 :
                        divisor = 4;
                        Log.d("Divisor", " "+ divisor);
                        break;
                }

            } });





   calculate.setOnClickListener(v -> {
    if (TextUtils.isEmpty(inverterLoad.getText()) || TextUtils.isEmpty(batteryNumber.getText())){
        Toast.makeText(getActivity(), "Fill in Values", Toast.LENGTH_SHORT).show();
        return;
    }else {

        int totalBatteries = Integer.parseInt(batteryNumber.getText().toString());
        int batteryAmps = Calculator.ampsConverter(totalBatteries, divisor, batteryAmp);
        int load = Integer.parseInt(inverterLoad.getText().toString());
        time2Charge = Calculator.timeToCharge(batteryAmps, invertVoltage);
        backUpTime = Calculator.batteryBackupTime(load, invertVoltage, batteryAmps);

        waveLoadingView.setTopTitle("Charge Time: " + Calculator.formatDecimal(time2Charge) + "hrs");
        waveLoadingView.setBottomTitle("BackUp Time: " + Calculator.formatDecimal(backUpTime) + "hrs");
    }

   });

    }








}

File: app/src/main/java/com/sofrosyn/a6gnlsolarcalculator/fragments/InverterRatingFragment.kt
package com.sofrosyn.a6gnlsolarcalculator.fragments

import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation
import androidx.navigation.findNavController
import com.google.android.material.button.MaterialButton
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textview.MaterialTextView
import com.sofrosyn.a6gnlsolarcalculator.R
import com.sofrosyn.a6gnlsolarcalculator.Utils.StateManager
import com.sofrosyn.a6gnlsolarcalculator.logic.Calculator.ampsConverter
import com.sofrosyn.a6gnlsolarcalculator.logic.Calculator.batteryBackupTime
import com.sofrosyn.a6gnlsolarcalculator.logic.Calculator.formatDecimal
import com.sofrosyn.a6gnlsolarcalculator.logic.Calculator.loadInKva
import com.sofrosyn.a6gnlsolarcalculator.logic.Calculator.timeToCharge
import java.util.*

import java.util.stream.IntStream

class InverterRatingFragment : Fragment() ,View.OnClickListener {
    private var inverterLoad: TextInputEditText? = null
    private var batteryNumber: TextInputEditText? = null
    private var estInverterRating: MaterialTextView? = null

    private var batteryChips: ChipGroup? = null
    private var inverterVoltageChip: ChipGroup? = null
    //private var view: View? = null
    private var inverterNext: MaterialButton? = null
    private var calculate: MaterialButton? = null
    private var acceptedBatteryCount: ArrayList<Int>? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
      val  view = inflater.inflate(R.layout.fragment_inverter_rating, container, false)
        initViews()
        initActions()
        return view
    }

    private fun initViews() {
        inverterLoad = view?.findViewById(R.id.fragment_inverter_load)
        // batteryLoad = view.findViewById(R.id.fragment_battery_load);
        estInverterRating = view?.findViewById(R.id.fragment_inverter_load_text)
        inverterVoltageChip = view?.findViewById(R.id.inverter_voltage)
        inverterNext = view?.findViewById(R.id.inverter_next)
        batteryNumber = view?.findViewById(R.id.fragment_inverter_battery_no)
        calculate = view?.findViewById(R.id.inverter_calculate)
//        waveLoadingView = view?.findViewById<>(R.id.waveLoadingView)
        batteryChips = view?.findViewById(R.id.chips_battery)
        acceptedBatteryCount = ArrayList()
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private fun initActions() {
        inverterLoad!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (start > 1) {
                    ratingKva = loadInKva(s.toString().toDouble())
                    estInverterRating!!.text = String.format("Your Estimated Inverter Rating is [%sKva] ", ratingKva)
                } else {
                    estInverterRating!!.text = "Your Estimated Inverter Rating  [Kva] "
  //                  waveLoadingView!!.centerTitle = "Load"
                }
            }

            override fun afterTextChanged(s: Editable) {
    //            waveLoadingView!!.progressValue = ratingKva.toInt() * 10
                estInverterRating!!.setTextColor(resources.getColor(R.color.colorPrimaryDark))
      //          waveLoadingView!!.centerTitle = "Inverter: " + ratingKva + "Kva"
            }
        })
        inverterNext!!.setOnClickListener { v: View? ->
            val load = inverterLoad!!.text.toString().toInt()
            val batteryCount = batteryNumber!!.text.toString().toInt()
            StateManager.getInstance(activity)
                    .Inverter(load, ratingKva.toFloat(), backUpTime.toFloat(), invertVoltage, time2Charge.toFloat(), batteryAmp, invertVoltage, batteryCount)
            //Navigation.findNavController(view?).navigate(R.id.action_inverterRatingFragment_to_batteryFragment)
           // it.findNavController().navigate(EmployeesListFragmentDirections.actionEmployeesListFragmentToAddEmployeeFragment())



        }
        batteryChips!!.setOnCheckedChangeListener { group: ChipGroup, checkedId: Int ->
            val chip: Chip = group.findViewById(checkedId)
            if (chip != null) {
                batteryAmp = chip.text.toString().toInt()
                Toast.makeText(activity, "Battery Amp: $batteryAmp", Toast.LENGTH_SHORT).show()
            }
        }
        inverterVoltageChip!!.setOnCheckedChangeListener { group: ChipGroup, checkedId: Int ->
            val chip: Chip = group.findViewById(checkedId)
            if (chip != null) {
                invertVoltage = chip.text.toString().toInt()
                Toast.makeText(activity, "Inverter Voltage: $invertVoltage", Toast.LENGTH_SHORT).show()
                when (invertVoltage) {
                    12 -> {
                        divisor = 1
                        acceptedBatteryCount!!.clear()
                        acceptedBatteryCount!!.add(IntStream.rangeClosed(1, 10).iterator().nextInt())
                        Log.d("Divisor", " $divisor")
                    }
                    24 -> {
                        divisor = 2
                        acceptedBatteryCount!!.clear()
                        IntStream.range(2, 20).filter { x: Int -> x % 2 == 0 }.forEach { x: Int -> acceptedBatteryCount!!.add(x) }
                        Log.d("Divisor", " $divisor")
                    }
                    48 -> {
                        divisor = 4
                        acceptedBatteryCount!!.clear()
                        IntStream.range(4, 20).filter { x: Int -> x % 4 == 0 }.forEach { x: Int -> acceptedBatteryCount!!.add(x) }
                        Log.d("Divisor", " $divisor")
                    }
                }
            }
        }
        calculate!!.setOnClickListener {
            if (!validate()) {
                return@setOnClickListener
            } else {
                val totalBatteries = batteryNumber!!.text.toString().toInt()
                val batteryAmps = ampsConverter(totalBatteries, divisor, batteryAmp)
                val load = inverterLoad!!.text.toString().toInt()
                time2Charge = timeToCharge(batteryAmps.toFloat(), invertVoltage.toFloat()).toDouble()
                backUpTime = batteryBackupTime(load.toFloat(), invertVoltage.toFloat(), batteryAmps.toFloat()).toDouble()
        //        waveLoadingView!!.topTitle = "Charge Time: " + formatDecimal(time2Charge) + "hrs"
          //      waveLoadingView!!.bottomTitle = "BackUp Time: " + formatDecimal(backUpTime) + "hrs"
            }
        }
    }

    private fun validate(): Boolean {
        if (TextUtils.isEmpty(inverterLoad!!.text) || TextUtils.isEmpty(batteryNumber!!.text)) {
            Toast.makeText(activity, "Fill in Values", Toast.LENGTH_SHORT).show()
            return false
        }
        if (!acceptedBatteryCount!!.contains(batteryNumber!!.text.toString().toInt())) {
            Toast.makeText(activity, "Accepted Battery Count $acceptedBatteryCount", Toast.LENGTH_LONG).show()
            return false
        }
        return true
    }

    companion object {
        private var batteryAmp = 100
        private var invertVoltage = 12
        private var divisor = 1
        private var ratingKva = 0.0
        private var time2Charge = 0.0
        private var backUpTime = 0.0
    }

    override fun onClick(view: View) {
        val action =
                InverterRatingFragmentDirections.actionInverterRatingFragmentToBatteryFragment()
        view.findNavController().navigate(action)
    }

}